Wikipedia/UI-V5/WMFSaveButtonController.m

```diff
@@ -0,0 +1,83 @@
+//  Created by Monte Hurd on 8/19/15.2//  Copyright (c) 2015 Wikimedia Foundation. Provided under MIT-style license; please copy and modify!
+
+#import "WMFSaveButtonController.h"
+#import "MWKSavedPageList.h"
+#import "MWKArticle.h"
+#import "SessionSingleton.h"
+#import "MWKUserDataStore.h"
+
+@interface WMFSaveButtonController ()
+
+@property (strong, nonatomic) UIButton* button;
+
+@end
+
+@implementation WMFSaveButtonController
+
+- (instancetype)initWithButton:(UIButton*)button
+                         title:(MWKTitle*)title {
+    self = [super init];
+    if (self) {
+        self.button = button;
+        self.title  = title;
+        [self observeSavedPages];
+        [self updateSavedButtonState];
+    }
+    return self;
+}
+
+- (MWKSavedPageList*)savedPageList {
+    return [SessionSingleton sharedInstance].userDataStore.savedPageList;
```

Comment:
@montehurd **TL;DR;** I don't think that's sufficient reason to abandon DI. I think we should continue to pass (note, not create) saved page lists.  Synchronizing saved page lists is a separate issue which we can resolve separately if needed.

---

I also considered this vulnerability, but I consider it to be an edge case.  You could make the same argument for every other `MWKSavedPageList`-containing class in the app, but my opinion is that the architectural benefits of loose coupling outweigh the potential problems—which we have yet to encounter in practice AFAIK.*

From what I can see we have two problems:

1. How do we ensure saved pages is always up to date
2. How do we acquire a class's dependencies?

They are not orthogonal, but I consider the latter to be solved through the use of dependency injection.  The former, OTOH, is an unfortunate legacy issue which even a singleton doesn't solve.**  If we want, we can synchronize saved page lists rather easily:

- Post a notification from the data store (which is **the** way to save them\***)
- Observe that notification in saved page list
- Check if the list that was saved is equal to `self`
- If not, reset in-memory storage, which is lazily reloaded from disk on next access

\* Issues w/ the lists not updating is a symptom of poorly factored logic that relied on a single, multi-responsibility object to make all necessary updates, and should be fixed once we factor that logic out into a separate mechanism which doesn't rely on any one code path.
\** Creating and saving another saved page list could overwrite the data and knock everything else out of sync until next save.
\*** Although not the only way...